### PR TITLE
Good eggs 0 & 1

### PR DESCRIPTION
**Pull request by mgasner**

### CODE REVIEW

File: pages/blog/good-eggs-1.mdx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+If our type checks succeed, we know we can move ahead with data ingest. We conditionally execute solids that write the data to S3, then to Snowflake. Then, we trigger downstream dbt models, dbt tests, and Jupyter notebooks. 
+
+For us, failing before bad data is ingested to the warehouse is a big win. It’s a little bit like the value provided by database schema constraints, but in a system (Google Sheets) with no concept of schema or typing. You can’t commit strings to an integer column in a database, and you shouldn’t be able to load a form with values of the wrong type either.
+
+We recognized the value of being able to impose strong typing at will on any of our inputs or outputs when we first started to evaluate data orchestrators, and this was one of the factors in our choice of Dagster over Airflow.
+
+<p align="center">
+  <img
+    src="/posts/good-eggs-1/slack_failure.png"
```
Comment (schrockn) [inline pages/blog/good-eggs-1.mdx:96]: this image isn't showing up

File: pages/blog/good-eggs-1.mdx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+If our type checks fail, we conditionally execute solids that notify us in Slack and via email of the ingest failures, so we can take corrective action quickly.
+
+<p align="center">
+  <img
+    src="/posts/good-eggs-1/validation_email.png"
+    alt="Good Eggs Validation Failure Email"
+    width="600px"
+  />
+</p>
```
Comment (schrockn) [inline pages/blog/good-eggs-1.mdx:86]: this image is not showing up

File: pages/blog/good-eggs-1.mdx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+We also yield structured metadata as part of our type check process. Being able to view this metadata in the Dagster logs before ingestion has real benefits for the human in the loop. For example, we yield structured metadata with the head and tail of the data in a Google Sheet. This is viewable directly in Dagit and is often enough for a human to diagnose a data quality problem.
+
+<p align="center">
+  <img
+    src="/posts/good-eggs-1/tail_zoom_blur_new.gif"
+    alt="Good Eggs Data Frame Tail"
+    width="600px"
+  />
+</p>
```
Comment (schrockn) [inline pages/blog/good-eggs-1.mdx:112]: this image also not working

**Review by schrockn — CHANGES_REQUESTED** [review]

File: pages/blog/good-eggs-0.mdx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+First, we’ll talk about how Dagster’s support for custom data types helped us achieve better correctness and reliability in our data ingest process, which meant less downstream breakage and better recovery when bad data made it through.
+
+Then, we’ll discuss how Dagster’s support for structured metadata enabled users with very different backgrounds to self-serve when trying to connect data assets to the computations that produced them, reducing load on the data platform team.
+
+We’ll talk about how we use Dagster to manage the tech debt we incur in other tools and cut our spend on platform components.
+
+We’ll discuss how we deployed Dagster on our custom PaaS, and how its APIs let us integrate at the right level of abstraction.
+
+Finally, we’ll talk about the improvements to the software engineering lifecycle of our data platform that have grown out of Dagster’s ground-up commitment to testability.
```
[the remark this answers is not in the document]
Comment (mgasner) [reply pages/blog/good-eggs-0.mdx:52]: ok

File: pages/blog/good-eggs-1.mdx [hunk 5]
```diff
@@ -0,0 +1,116 @@
+---
+layout: 'BlogPost'
+status: 'published'
+title: 'Good Data at Good Eggs: Correctness and reliability for data infrastructure'
+excerpt: 'Dagster’s support for custom data types helped us achieve better correctness and reliability in our data ingest process, which meant less downstream breakage and better recovery when bad data made it through.'
+date: '2020-10-01'
+authors: ['davidjwallace']
+---
+
+_This is the second in a series of posts about our experience putting Dagster into production at Good Eggs._
```
[the remark this answers is not in the document]
Comment (mgasner) [reply pages/blog/good-eggs-1.mdx:10]: yep

File: pages/blog/good-eggs-0.mdx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+Adopting Dagster was a catalyst for the transformation of our data platform team. We hope our experience is encouraging to other teams facing similar challenges and opportunities.
```
Comment (schrockn) [inline pages/blog/good-eggs-0.mdx:54]: Could we specify *what* the transformation was? transformation of our data platform team from XXXX to YYYY

File: pages/blog/good-eggs-1.mdx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  />
+</p>
+
+Let’s look at how we manage our data ingest from Google Sheets using Dagster.
+
+Each node in the pipeline here is what Dagster calls a *solid*, a functional unit of computation that products and consumes data assets.
+
+In our pipeline, a solid reads a range of values from a Google Sheet, and then reshapes it as a Pandas data frame with given column types.
+
+We then type check the data against a custom data frame type as a way of running quality checks on the data. In Dagster, type checks can run arbitrary code and associate their results with arbitrary structured metadata. Here, we use a type factory from the `dagster_pandas` library to construct a custom data frame, which runs a set of checks on the data frame and its column. This process guarantees invariants, like that values for certain columns are non-null or belong to certain sets of categorical values. We also inject structued metadata (generated by our own `compute_dataframe_summary_statistics` function) so that summary statistics about the data appear in Dagit and in longitudinal views.
```
[the remark this answers is not in the document]
Comment (mgasner) [reply pages/blog/good-eggs-1.mdx:30]: ok

File: pages/blog/good-eggs-1.mdx (hunk 2 above)
Comment (mgasner) [reply pages/blog/good-eggs-1.mdx:86]: fixed

File: pages/blog/good-eggs-1.mdx (hunk 3 above)
Comment (mgasner) [reply pages/blog/good-eggs-1.mdx:112]: fixed

File: pages/blog/good-eggs-1.mdx [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+We recognized the value of being able to impose strong typing at will on any of our inputs or outputs when we first started to evaluate data orchestrators, and this was one of the factors in our choice of Dagster over Airflow.
+
+<p align="center">
+  <img
+    src="/posts/good-eggs-1/slack_failure.png"
+    alt="Good Eggs Slack Failure"
+    width="600px"
+  />
+</p>
```
[the remark this answers is not in the document]
Comment (mgasner) [reply pages/blog/good-eggs-1.mdx:100]: rn we have "If our type checks fail, we conditionally execute solids that notify us in Slack and via email of the ingest failures, so we can take corrective action quickly."

File: pages/blog/good-eggs-0.mdx (hunk 6 above)
Comment (mgasner) [reply pages/blog/good-eggs-0.mdx:54]: defer to @dwallace0723

File: pages/blog/good-eggs-1.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (schrockn) [reply pages/blog/good-eggs-1.mdx:100]: Yes but then the slack screenshot is totally out of context. That's the issue.

I think you can explain before the email is that is sent to select users (e.g. the ops team) and is part of the business logic of the pipeline.

For *any* operational failure, they push these messages to slack, where they can quickly and easily diagnose production issues.

[the remark this answers is not in the document]
Comment (mgasner) [reply pages/blog/good-eggs-1.mdx:96]: k

**Review by schrockn — APPROVED** [review]

File: pages/blog/good-eggs-1.mdx [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+       PandasColumn.datetime_column(
+           name="date",
+           min_datetime=START_DATE,
+           max_datetime=datetime.utcnow(),
+           non_nullable=True,
+       ),
+       PandasColumn.string_column(name="team", non_nullable=True),
+       PandasColumn.numeric_column(name="hours", min_value=0, non_nullable=True),
+       PandasColumn.categorical_column(
+           name="location", categories={...}, non_nullable=True
```
Comment (dwallace0723) [inline pages/blog/good-eggs-1.mdx:59]:
```suggestion
           name="location", categories={"sfbay", "MAR1"}, non_nullable=True
```

**Review by dwallace0723 — APPROVED** [review]
looks great 💯 👏  Thanks for all the hard work - it's been such a pleasure collaborating with y'all on this!!

**Pull request merged**